cpp/src/arrow/array/util.cc

```diff
@@ -511,6 +511,9 @@ class NullArrayFactory {
   }
 
   Status Visit(const ExtensionType& type) {
+    std::vector<std::shared_ptr<ArrayData>> child_data(type.storage_type()->num_fields());
+    out_ =
+        ArrayData::Make(type.storage_type(), length_, {buffer_}, child_data, length_, 0);
```

Comment:
Looks like this causes the Python tests to fail. Should `out_` have the storage type instead of the extension type? It seems just `out_->child_data.resize(type.storage_type()->num_fields())` should be enough here.

Reply:
Ah, thanks, you are correct.  I switched to that and added a check for the array's type to the unit test.